staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go

```diff
@@ -238,43 +325,272 @@ func newAuthenticator(opts Options, initVerifier func(ctx context.Context, a *Au
+// TODO: Support separate client IDs per different issuer/claim.
+func newClaimResolver(ctx context.Context, issuersPerClaim map[string][]string, t *http.Transport, config *oidc.Config) *claimResolver {
+	cp := &claimResolver{claimInfos: map[string]claimInfo{}, t: t}
+	for c, issuers := range issuersPerClaim {
+		verifierPerIssuer := map[string]*asyncIDTokenVerifier{}
+		for _, iss := range issuers {
+			glog.V(10).Infof("adding claim info: %v", iss)
+			verifierPerIssuer[iss] = newAsyncIDTokenVerifier(ctx, config, iss)
+		}
+		ci := claimInfo{c, verifierPerIssuer}
+		cp.claimInfos[c] = ci
+		glog.V(10).Infof("adding claim info: %+v", ci)
+	}
+	return cp
+}
+
+// Verifier returns either the verifier for the specified issuer, or error.
+func (p *claimResolver) Verifier(claim, iss string) (*oidc.IDTokenVerifier, error) {
+	ci, ok := p.claimInfos[claim]
+	if !ok {
+		return nil, fmt.Errorf("no claim info for claim: %q, issuer: %q", claim, iss)
+	}
+	av := ci.verifierPerIssuer[iss]
+	if av == nil {
+		return nil, fmt.Errorf("no verifier for claim: %q, issuer: %q", claim, iss)
+	}
+	v := av.verifier()
+	if v == nil {
+		return nil, fmt.Errorf("verifier not initialized for claim: %q, issuer: %q", claim, iss)
+	}
+	glog.V(10).Infof("Got verifier for: claim: %q, iss: %q: %+v", claim, iss, v)
+	return v, nil
+}
+
+// expand extracts the distributed claims from claim names and claim sources.
+// The extracted claim value is pulled up into the supplied claims.
+//
+// Distributed claims are of the form as seen below, and are defined in the
+// OIDC Connect Core 1.0, section 5.6.2.
+// See: https://openid.net/specs/openid-connect-core-1_0.html#AggregatedDistributedClaims
+//
+// {
+//   ... (other normal claims)...
+//   "_claim_names": {
+//     "groups": "src1"
+//   },
+//   "_claim_sources": {
+//     "src1": {
+//       "endpoint": "https://www.example.com",
+//       "access_token": "f005ba11"
+//     },
+//   },
+// }
+func (r *claimResolver) expand(c claims) error {
+	const (
+		// The claim containing a map of endpoint references per claim.
+		// OIDC Connect Core 1.0, section 5.6.2.
+		claimNamesKey = "_claim_names"
+		// The claim containing endpoint specifications.
+		// OIDC Connect Core 1.0, section 5.6.2.
+		claimSourcesKey = "_claim_sources"
+	)
+
+	if glog.V(11) {
+		glog.Infof("initial claim set: %v", c)
+	}
+	defer func() {
+		if glog.V(10) {
+			glog.Infof("final claim set: %v", c)
+		}
+	}()
+	names, ok := c[claimNamesKey]
+	if !ok {
+		// No _claim_names, no keys to look up.
+		return nil
+	}
+
+	claimToSource := map[string]string{}
+	if err := json.Unmarshal([]byte(names), &claimToSource); err != nil {
+		return fmt.Errorf("oidc: error parsing distributed claim names: %v", err)
+	}
+
+	var rawSources json.RawMessage
+	rawSources, ok = c[claimSourcesKey]
+	if !ok {
+		// Having _claim_names claim,  but no _claim_sources is not an expected
+		// state.
+		return fmt.Errorf("oidc: no claim sources")
+	}
+
+	var sources claims
```

Comment:
can this be `map[string]endpoint` so you don't have to unmarshal endpoints in the loop below?

Reply:
Lo and behold, it can!

Done. :)